qemu/tests/slof_boot.py

```diff
@@ -0,0 +1,117 @@
+from virttest import utils_net
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Verify SLOF info by booting guest with different devices and options.
+    Step:
+        1. Boot a guest with different storage devices and others options.
+            Including storage devices:
+            a. virtio-blk
+            b. virtio-blk with dataplane
+            c. virtio-scsi
+            d. vritio-scsi with dataplane
+            e. spapr-vscsi
+            g. virtio-scsi disk behind pci-bridge
+            h. virtio-blk-pci disk behind pci-bridge
+            i. virtio-blk with block size
+            j. virtio-scsi with block size
+        2. Check if any error info in output of SLOF during booting.
+        3. Check if booted from specified storage device successfully by
+            reading "Trying to load:  from: /xxx/xxx/xxx ..." from output of SLOF.
+        4. Check if login guest successfully.
+        5. Guest could ping external host ip.
+    :param test: Qemu test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    def _get_pci_bridge_addr(id):
+        dev_info_list = re.findall(r'\s+-device pci-bridge,\S+\s+',
+                                   vm.qemu_command)
+        for dev_info in dev_info_list:
+            if ('id=' + id) in dev_info:
+                return re.search(r'(addr)=(\w+)', dev_info).group(2)
+
+    def _verify_boot_status(boot_dev):
+        dev_params = params.object_params(boot_dev)
+        child_addr = dev_params.get('child_addr')
+        sub_child_addr = dev_params.get('sub_child_addr', None)
+        parent_bus = dev_params.get('parent_bus')
+        child_bus = dev_params.get('child_bus')
+        if child_bus == 'pci-bridge':
+            pci_bus_id = params.get('pci_bus_image1', None)
+            child_addr = _get_pci_bridge_addr(pci_bus_id)
+        if sub_child_addr:
+            verify_info = ('Verify whether booting from specified %s device (%s@%s).'
+                           % (boot_dev, child_addr, sub_child_addr))
+            fail_info = ('Failed to boot from %s device (%s@%s).'
+                         % (boot_dev, child_addr, sub_child_addr))
+            ret_info = ('Boot from %s device (%s@%s) successfully.'
+                        % (boot_dev, child_addr, sub_child_addr))
+        else:
+            verify_info = ('Verify whether booting from specified %s device (@%s).'
+                           % (boot_dev, child_addr))
+            fail_info = ('Failed to boot from %s device (@%s).'
+                         % (boot_dev, child_addr))
+            ret_info = ('Boot from %s device (@%s) successfully.'
+                        % (boot_dev, child_addr))
+
+        logging.info(verify_info)
+
+        if not slof.verify_boot_device(vm, parent_bus, child_bus,
+                                       child_addr, sub_child_addr):
+            test.fail(fail_info)
+        else:
```

Comment:
`else:` can be omitted and unintended the next line

Reply:
Updated , please check. thanks.